### PR TITLE
WIP [DO_NOT_MERGE]: Initial addition of S2I image extended tests

### PR DESCRIPTION
**Pull request by mfojtik**
This test replaces our bash tests for SCL usage and S2I usage. Can be re-used for other images.

### CODE REVIEW

**Comment by mfojtik** [comment]
@mnagy @gabemontero @jhadvig @bparees  (this is just a prototype ;-)

**Comment by mfojtik** [comment]
Sample output: https://gist.github.com/mfojtik/9836d118ee503dbbd0ff

File: test/extended/default_builds_test.go [hunk 1]
```diff
@@ -59,14 +59,9 @@ var _ = Describe("STI build with .sti/environment file", func() {
...
-			outputPath := filepath.Join(testContext.OutputDir, oc.Namespace()+"-sample-pod.json")
-			pod := exutil.CreatePodForImage(imageName)
-			err = exutil.WriteObjectToFile(pod, outputPath)
-			Expect(err).NotTo(HaveOccurred())
-
-			By(fmt.Sprintf("calling oc create -f %q", outputPath))
-			err = oc.Run("create").Args("-f", outputPath).Execute()
+			By(fmt.Sprintf("create a new pod for %q", imageName))
+			pod := exutil.GetPodForImage(imageName)
+			_, err := oc.KubeREST().Pods(oc.Namespace()).Create(pod)
```
Comment (mnagy) [inline test/extended/default_builds_test.go]: test/extended/default_builds_test.go:64: no new variables on left side of :=

File: test/extended/s2i_ruby_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			//"22-rhel7":   "ruby 2.2.2",
+		}
+	)
+
+	for v, expected := range versions {
+
+		Describe(fmt.Sprintf("SCL usage of ruby-%s", v), func() {
+			It("should provide usage", func() {
+				oc.SetOutputDir(testContext.OutputDir)
+				name := exutil.MustGetImageName(oc.REST().ImageStreams("openshift"), "ruby", "latest")
```
Comment (mfojtik) [inline test/extended/s2i_ruby_test.go]: Need to brainstorm better name for this ;-)

File: test/extended/s2i_ruby_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+var _ = Describe("OpenShift Ruby S2I image", func() {
+	defer GinkgoRecover()
+	var (
+		oc       = exutil.NewCLI("mysql-create", kubeConfigPath())
+		versions = map[string]string{
+			"20-centos7": "ruby 2.0.0",
+			//"20-rhel7":   "ruby 2.0.0",
+			//"22-centos7": "ruby 2.2.2",
```
Comment (mfojtik) [inline test/extended/s2i_ruby_test.go]: Just a note for the future @mfojtik, this needs to be enabled once the 22 image is available on Hub.

File: test/extended/s2i_ruby_test.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+					Name:  "test",
+					Image: name,
+				})
+				oc.KubeFramework().TestContainerOutput("ruby-"+v, pod, 0, []string{"Sample invocation"})
+			})
+		})
+
+		Describe(fmt.Sprintf("SCL usage of ruby-%s", v), func() {
+			It("should allow various invocations of binary", func() {
+				oc.SetOutputDir(testContext.OutputDir)
```
Comment (mfojtik) [inline test/extended/s2i_ruby_test.go]: This should be moved to Before() helper

File: test/extended/s2i_ruby_test.go (hunk 3 above)
Comment (bparees) [inline test/extended/s2i_ruby_test.go]: it's available.

File: test/extended/s2i_ruby_test.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	. "github.com/onsi/gomega"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = Describe("OpenShift Ruby S2I image", func() {
+	defer GinkgoRecover()
+	var (
+		oc       = exutil.NewCLI("mysql-create", kubeConfigPath())
```
Comment (bparees) [inline test/extended/s2i_ruby_test.go]: mysql-create?

File: test/extended/s2i_ruby_test.go (hunk 2 above)
Comment (bparees) [inline test/extended/s2i_ruby_test.go]: what registered this imagestream?  how do we know it's referencing the image we want instead of something it found on dockerhub?  (this will be a bigger problem with v2 registries).  

I think the GetPodForContainer should use a straight docker pullspec that we know can be resolved locally if the image was built on this machine.

Comment (mfojtik) [inline test/extended/s2i_ruby_test.go]: @bparees it is using the imagestreams from the `openshift` namespace that we import when we start the server. I assume that the 'latest' ImageStreamTag will be used which in combination of using `--latest-images` option when starting the server will grab the "latest" Docker image available locally.

File: test/extended/s2i_ruby_test.go (hunk 5 above)
Comment (mfojtik) [inline test/extended/s2i_ruby_test.go]: hups. copy&paste

File: test/extended/s2i_ruby_test.go (hunk 2 above)
Comment (mfojtik) [inline test/extended/s2i_ruby_test.go]: @bparees also I do wanna test this the same way as our users will use it (which is through IS).

Comment (bparees) [inline test/extended/s2i_ruby_test.go]: @mfojtik I don't think that's going to do what you want, or not guaranteed anyway.

--latest-images has no bearing on what builder image (eg ruby-20-centos7)  the build flows use.

The ImageStream is going to import from the registry when the imagestream is created (and is that the rhel7 or centos imagestream you're creating?  or do you run the test twice, once for each imagestream?) which means it's going to set the imageid to whatever is in the registry at the time the imagestream was created.

The build, when run, is going to resolve the ImageStreamTag to a pullspec which in this case might end in :latest in which case yes s2i will default to using the existing image if present, but at some point the ImageStreamTag might point to a pullspec which instead ends in a :imageid (v2 registry style) in which case the local image you just built will be totally ignored.

and all that assumes we never set the BuildConfig forcePull setting to true.

some further discussion w/ andy on irc:

agoldste: if my imagestream points to a v2 registry
(10:10:22 AM) bparees: agoldste: the dockerImageReference for a given ImageStreamTag will be of the form something/image@id right?
(10:29:09 AM) bparees: agoldste: ok.  so was my description of the behavior wrong?
(10:29:33 AM) agoldste: bparees: it's correct, but we don't support importing from v2 registries yet
(10:29:50 AM) agoldste: i.e. spec.dockerImageRepository = someV2registry/some/repo won't import any tags
(10:30:09 AM) bparees: agoldste: ok.....what if the imagestream declares some tags in its spec?  will we import those?
(10:30:37 AM) agoldste: we need to rework pruning so the registry owns pruning its data (instead of the oadm prune images command making http prune requests to the registry), then we need to bump the docker/distribution code, then we need to modify the image import controller to support v2 imports
(10:30:53 AM) agoldste: we can't import any data from a v2 registry right now
(10:31:06 AM) agoldste: other than the direct integration w/have with our internal registry

So basically once we move to v2 importing this logic will break because it'll try to run the imageid that exists in the registry, not the one you just built locally.

Comment (mfojtik) [inline test/extended/s2i_ruby_test.go]: then we should just use DockerImage directly instead of IS

Comment (bparees) [inline test/extended/s2i_ruby_test.go]: sounds reasonable to me.

**Comment by mfojtik** [comment]
See: https://github.com/openshift/origin/pull/4286
